src/client/datascience/jupyter/kernels/kernelConnectionWrapper.ts

```diff
@@ -44,84 +44,124 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
     public readonly unhandledMessage = new Signal<this, IMessage<MessageType>>(this);
     public readonly anyMessage = new Signal<this, Kernel.IAnyMessageArgs>(this);
     public get serverSettings() {
-        return this.kernelConnection.serverSettings;
+        return (this.kernelConnection || this._previousKernelConnection).serverSettings;
     }
     public readonly disposed = new Signal<this, void>(this);
+    private _kernelConnection!: Kernel.IKernelConnection;
+    private _previousKernelConnection!: Kernel.IKernelConnection;
+    // private _isRestarting?: boolean;
+    private get kernelConnection(): undefined | Kernel.IKernelConnection {
+        if (this.kernel.session?.kernel === this._kernelConnection) {
+            return this._kernelConnection;
+        }
+        this.stopHandlingKernelMessages(this._kernelConnection);
```

Comment:
If the previous kernel connection isn't the same as the current, then kernel was restarted and we have a new connection.
Hence remove event handlers associated with the previous connections,